workspaces/diff-engine/cli/src/main.rs

```diff
@@ -106,57 +109,105 @@ fn main() {
       }
     });
 
+    let mut diff_tasks: Vec<JoinHandle<()>> = Vec::with_capacity(diff_queue_size);
+
+    tokio::pin!(results_manager);
+
     dbg!("waiting for next interaction");
-    while let Some(interaction_json_result) = interaction_lines.next().await {
-      //dbg!("got next interaction");
-      let diff_permits = diff_scheduling_permits.clone();
-      let projection = spec_projection.clone();
-      let mut results_sender = results_sender.clone();
-      //dbg!("waiting for permit");
-      let diff_task_permit = diff_permits.acquire_owned().await;
-      //dbg!("got permit");
-
-      tokio::spawn(async move {
-        let diff_comp =
-          tokio::task::spawn_blocking::<_, Option<(Vec<InteractionDiffResult>, Tags)>>(move || {
-            let interaction_json =
-              interaction_json_result.expect("can read interaction json line from stdin");
-            let TaggedInput(interaction, tags): TaggedInput<HttpInteraction> =
-              match serde_json::from_str(&interaction_json) {
-                Ok(tagged_interaction) => tagged_interaction,
-                Err(parse_error) => {
-                  eprintln!("could not parse interaction json: {}", parse_error);
-                  return None;
-                }
-              };
 
-            Some((diff_interaction(&projection, interaction), tags))
-          });
-        //dbg!("waiting for results");
-        let results = diff_comp
-          .await
-          .expect("diffing of interaction should be successful");
-        //dbg!("got results");
-
-        if let Some((results, tags)) = results {
-          for result in results {
-            //dbg!(&result);
-            if let Err(_) = results_sender
-              .send(ResultContainer::from((result, &tags)))
+    loop {
+      let next_action = tokio::select! {
```

Comment:
This running of `tokio::select!` in a loop is the core change of how we're diffing interactions one by one. We're doing multiple things asynchronously (reading interactions from `stdin`, computing the diff and writing results to `stdout`), this allows us to handle the next thing that's ready. Tokio polls these future at random, and in the case where that thing is happening on the same (main) thread (like the reading of interactions), the polling actually causes each future to progress. Spawning a task through Tokio causes progress to be made regardless.